[dynamo][eval frame] Make CacheEntry a PyObject

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107496
* __->__ #107405
* #107117
* #106917


This PR makes CacheEntry a PyObject. This is prep PR for cache size changes. As CacheEntry is a py object, we can now traverse the linked list in Python and write cache size policies. It was possible to do in C, but Python is just easier to iterate upon. We call convert_frame only when we (re)compile, so a small bump in latency going from C to Python is acceptable here.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

anijain2305 committed:
[dynamo][eval frame] Make CacheEntry a PyObject

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,49 @@ typedef struct cache_entry {
 
+#define Py_NONE_CACHE_ENTRY (CacheEntry*)Py_None
+
+static void destroy_cache_entry(CacheEntry* e);
```
You don't need this anymore right? You can just Py_DECREF now

anijain2305 replied:
This is the destructor of CacheEntry object. This will be called when ref count is 0. So, we do need this.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,49 @@ typedef struct cache_entry {
+#define DECLARE_CACHE_ENTRY_ATTR(name) \
+static PyObject* CacheEntry_##name(CacheEntry* self, PyObject* _noargs) { \
+  PyObject* res = (PyObject*)self->name; \
+  Py_XINCREF(res); \
```
Py_INCREF, this must not be nullptr.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,49 @@ typedef struct cache_entry {
+  .tp_name = "torch._C.dynamo.eval_frame.CacheEntryWrapper",
+  .tp_basicsize = sizeof(CacheEntry),
+  .tp_itemsize = 0,
+  .tp_flags = Py_TPFLAGS_DEFAULT,
```
You probably want GC support on this type too

anijain2305 replied:
Do you mean we need to define `tp_traverse` etc for this type? https://docs.python.org/3/c-api/gcsupport.html#supporting-cycle-detection

ezyang replied:
Yes. GC support will be useful for dealing with the cycles induced by doubly linking, which you said you wanted to do later. It's not strictly required, but if you don't have it you have to manually break all cycles when you want to deallocate.

If you want to see an example here is one I did recently: https://github.com/pytorch/pytorch/pull/107438

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,49 @@ typedef struct cache_entry {
+  .tp_flags = Py_TPFLAGS_DEFAULT,
+  .tp_new = PyType_GenericNew,
+  .tp_dealloc = (destructor)destroy_cache_entry,
+  .tp_getset = CacheEntry_properties,
```
lol, the only reason you can use aggregate initialization here is because this is a C file lololol

ezyang replied:
(in C++ this is C++20, I believe)

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,49 @@ typedef struct cache_entry {
+  .tp_basicsize = sizeof(CacheEntry),
+  .tp_itemsize = 0,
+  .tp_flags = Py_TPFLAGS_DEFAULT,
+  .tp_new = PyType_GenericNew,
```
Using GenericNew is probably not the right call, since you have fields on cache_entry which need to be initialized first, so if you do it as you've done here, if you call a getter after you have constructed the object you will segfault because the fields are NULL.

anijain2305 replied:
Good point. Added a `tp_new` and `tp_init`.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -338,13 +378,16 @@ static CacheEntry* create_cache_entry(
 
 static void destroy_cache_entry(CacheEntry* e) {
-  if (e == NULL || e == SKIP_CODE) {
+  if (e == Py_NONE_CACHE_ENTRY) {
```
For the tp_dealloc, this case is impossible.

For all non-tp_dealloc use cases, you should be simply Py_DECREF now.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -338,13 +378,16 @@ static CacheEntry* create_cache_entry(
 }
 
 static void destroy_cache_entry(CacheEntry* e) {
-  if (e == NULL || e == SKIP_CODE) {
```
What happened to SKIP_CODE? Did you just get rid of it.

anijain2305 replied:
This is wrong on main branch. Our cache_entry is never SKIP_CODE. Our extra_state can be SKIP_CODE. I missed removing this in my last PR.

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"



This PR makes CacheEntry a PyObject. This is prep PR for cache size changes. As CacheEntry is a py object, we can now traverse the linked list in Python and write cache size policies. It was possible to do in C, but Python is just easier to iterate upon. We call convert_frame only when we (re)compile, so a small bump in latency going from C to Python is acceptable here.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"



This PR makes CacheEntry a PyObject. This is prep PR for cache size changes. As CacheEntry is a py object, we can now traverse the linked list in Python and write cache size policies. It was possible to do in C, but Python is just easier to iterate upon. We call convert_frame only when we (re)compile, so a small bump in latency going from C to Python is acceptable here.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"



This PR makes CacheEntry a PyObject. This is prep PR for cache size changes. As CacheEntry is a py object, we can now traverse the linked list in Python and write cache size policies. It was possible to do in C, but Python is just easier to iterate upon. We call convert_frame only when we (re)compile, so a small bump in latency going from C to Python is acceptable here.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,100 @@ typedef struct cache_entry {
+    self->check_fn = Py_None;
+    Py_INCREF(Py_None);
+    self->code = (PyCodeObject*)Py_None;
+    Py_INCREF(Py_None);
```
This should be Py_NONE_CACHE_ENTRY no? ;) (TBH, it's probably clearer to just cast Py_None to the appropriate type as necessary, instead of macro'ing it)

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,100 @@ typedef struct cache_entry {
+
+  if (check_fn) {
+    PyObject* tmp = self->check_fn;
+    self->check_fn = check_fn;
```
The parser here doesn't give you a new reference, it's a borrow. So if you want to stash it in self I think you have to `Py_INCREF` it.

anijain2305 replied:
So, I am confused here. The `check_fn` is originating from the `convert_frame.py`, and when `call_callback` returns the GuardedCode object to `eval_frame.c`, CacheEntry steals the reference. This particular line in the init function is just an intermediate stage in the plumbing, with finally extra scratch scape becoming the owner of the reference. Therefore, I did not INCREF.

Also, a few test cases fail if I do INCREF here. And I came up with above reasoning to satisfy myself. Does it make sense?

ezyang replied:
See https://github.com/pytorch/pytorch/pull/107405#discussion_r1299505687

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -327,24 +418,23 @@ static CacheEntry* create_cache_entry(
-  NULL_CHECK(e->code);
-  e->next = next;
+  PyObject* check_fn = PyObject_GetAttrString(guarded_code, "check_fn");
+  PyCodeObject* code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code");
```
This returns a new reference, and then Py_BuildValue also increments the refcount by one. But you only decref args. This means you are leaking.

You are manually managing memory in C; you are basically juggling chainsaws. Everything has to be done correctly. Ask @albanD for tips on how to check for memory leaks on a debug build of Python.

ezyang replied:
Note that you may have accidentally made everything work out because IIRC, the CacheEntry constructor wasn't properly incref'ing the arguments it was passed, so the refcounts might balance out here.

ezyang reviewed (approved):
Please fix the borrow/own accident. The rest of the changes are cosmetic. I did finish reviewing everything else.

anijain2305 committed:
Update on "[dynamo][eval frame] Make CacheEntry a PyObject"



This PR makes CacheEntry a PyObject. This is prep PR for cache size changes. As CacheEntry is a py object, we can now traverse the linked list in Python and write cache size policies. It was possible to do in C, but Python is just easier to iterate upon. We call convert_frame only when we (re)compile, so a small bump in latency going from C to Python is acceptable here.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -304,10 +305,101 @@ typedef struct cache_entry {
+    Py_XDECREF(tmp);
+  }
+
+  if (next) {
+    CacheEntry* tmp = self->next;
+    Py_INCREF(next);
+    self->next = next;
+    Py_XDECREF(tmp);
+  }
```
@ezyang One last review of these lines and the corresponding lines in `cache_entry_init`.

anijain2305 commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -327,24 +419,28 @@ static CacheEntry* create_cache_entry(
+  // equivalent to CacheEntry(check_fn, code, next) in Python
+  PyObject* args = Py_BuildValue("OOO", check_fn, code, next);
+  CacheEntry* e = (CacheEntry*)PyObject_CallObject((PyObject*)&CacheEntryType, args); // new reference
+  // CacheEntry e is the now the owner of old cachey entry next. This happens
+  // when we incref the next pointer in cache_entry_init.
+  Py_DECREF(next);
+  Py_DECREF(check_fn);
+  Py_DECREF(code);
+  Py_DECREF(args);
```
@ezyang One last review of these lines - https://github.com/pytorch/pytorch/pull/107405/files#r1299639440

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -f "unrelated CI failure"
